fix: rename define in plugin dependencies

julien opened the pull request:
This change is related to #139, instead of touching the minified files directly, here we use `@babel` to rename `define` to `__define_disabled__`: if someone can give me some pointers we could probably remove the entire `if`
but for the moment this fixes the issue.

wincent commented on support/copyPluginDependencies.js:
```diff
@@ -23,7 +27,25 @@ for (const [depType, deps] of Object.entries(depsConfig)) {
+
+				traverse(ast, {
+					enter(path) {
+						if (path.node.name === 'define') {
```
You could make this check a bit stricter... like:

```js
if (
    path.node.type === 'Identifier' &&
    path.node.name === 'define'
)
```

or better still:

```js
if (
    path.node.type === 'UnaryExpression' &&
    path.node.operator === 'typeof' &&
    path.node.argument.type === 'Identifier' &&
    path.node.argument.name === 'define'
)
```

The idea of using an AST-based transform is to make it precise, like a scalpel, so that you don't edit something that you don't mean to edit. So make it as precise as you can without it being brittle in the opposite sense.

To increase our confidence in this change, it would be good to grab a before and after diff just to "prove" it's not editing anything we don't want it to. Even as a once-off manual check pasted into the PR, that's still valuable

julien replied:
Thanks for the suggestion @wincent, that's what I meant by 

> if someone can give me some pointers we could probably remove the entire if
but for the moment this fixes the issue.

I'm sure there's a way to make this more robust, but this is already a good start.

Here are some screenshots of the `ckeditor/plugins/codemirror/vendors/vendors.js` **before** (on the left) and **after** (on the right) 


![](https://user-images.githubusercontent.com/5572/100324273-4b1b2680-2fc7-11eb-8e80-9a734171145f.png)

![](https://user-images.githubusercontent.com/5572/100324287-51110780-2fc7-11eb-8493-158192b164cd.png)

![](https://user-images.githubusercontent.com/5572/100324304-579f7f00-2fc7-11eb-9ed3-ed56721d55c1.png)

wincent reviewed (approved):
Looks reasonable to me but I'd still like to see a before-and-after diff pasted in here somewhere to document the effects of your change.

julien commented:
@wincent so fast to reply :joy:

julien commented:
btw, I'm not sure if there's a better way to show diffs, but if there is I'm interested

wincent commented:
> btw, I'm not sure if there's a better way to show diffs, but if there is I'm interested

I don't know if it's better but you can run `git diff` and use use a fenced code block of type `diff` to get it highlighted.

eg.

```
    ```diff
    diff --git a/greeting.txt b/greeting.txt
    new file mode 100644
    index 000000000000..b7c75fabee93
    --- /dev/null
    +++ b/greeting.txt
    @@ -0,0 +1 @@
    +hai
    ```
```

Renders as:

```diff
diff --git a/greeting.txt b/greeting.txt
new file mode 100644
index 000000000000..b7c75fabee93
--- /dev/null
+++ b/greeting.txt
@@ -0,0 +1 @@
+hai
```

julien commented:
@wincent I see ... In this case it creates a huge diff, I wonder if something is wrong

wincent commented:
One of the things about Babel is that it doesn't preserve whitespace, so you should do the diff with `git diff -b` (ignore changes in the amount of space) or `git diff -w` (ignore all whitespace changes).

I usually look at `-b` first, and if that isn't aggressive enough, `-w` (which may be too aggressive, which is why it's worth looking at both).

julien commented:
@wincent thanks for the tip.

Here's a diff 

```diff
diff --git a/ckeditor/plugins/codemirror/vendors/vendors.js b/ckeditor/plugins/codemirror/vendors/vendors.js
index dd1302231..87f950aff 100644
--- a/ckeditor/plugins/codemirror/vendors/vendors.js
+++ b/ckeditor/plugins/codemirror/vendors/vendors.js
@@ -1,4 +1,4 @@
-﻿(function(Q,da){"object"===typeof exports&&"undefined"!==typeof module?module.exports=da():"function"===typeof define&&define.amd?define(da):(Q=Q||self,Q.CodeMirror=da())})(this,function(){function Q(a){return new RegExp("(^|\\s)"+a+"(?:$|\\s)\\s*")}function da(a){for(var b=a.childNodes.length;0<b;--b)a.removeChild(a.firstChild);return a}function Z(a,b){return da(a).appendChild(b)}function r(a,b,c,d){a=document.createElement(a);c&&(a.className=c);d&&(a.style.cssText=d);if("string"==typeof b)a.appendChild(document.createTextNode(b));
+﻿(function(Q,da){"object"===typeof exports&&"undefined"!==typeof module?module.exports=da():"function"===__define_disabled__&&define.amd?define(da):(Q=Q||self,Q.CodeMirror=da())})(this,function(){function Q(a){return new RegExp("(^|\\s)"+a+"(?:$|\\s)\\s*")}function da(a){for(var b=a.childNodes.length;0<b;--b)a.removeChild(a.firstChild);return a}function Z(a,b){return da(a).appendChild(b)}function r(a,b,c,d){a=document.createElement(a);c&&(a.className=c);d&&(a.style.cssText=d);if("string"==typeof b)a.appendChild(document.createTextNode(b));
 else if(b)for(c=0;c<b.length;++c)a.appendChild(b[c]);return a}function y(a,b,c,d){a=r(a,b,c,d);a.setAttribute("role","presentation");return a}function A(a,b){3==b.nodeType&&(b=b.parentNode);if(a.contains)return a.contains(b);do if(11==b.nodeType&&(b=b.host),b==a)return!0;while(b=b.parentNode)}function m(){var a;try{a=document.activeElement}catch(b){a=document.body||null}for(;a&&a.shadowRoot&&a.shadowRoot.activeElement;)a=a.shadowRoot.activeElement;return a}function B(a,b){var c=a.className;Q(b).test(c)||
 (a.className+=(c?" ":"")+b)}function n(a,b){for(var c=a.split(" "),d=0;d<c.length;d++)c[d]&&!Q(c[d]).test(b)&&(b+=" "+c[d]);return b}function f(a){var b=Array.prototype.slice.call(arguments,1);return function(){return a.apply(null,b)}}function k(a,b,c){b||(b={});for(var d in a)!a.hasOwnProperty(d)||!1===c&&b.hasOwnProperty(d)||(b[d]=a[d]);return b}function I(a,b,c,d,e){null==b&&(b=a.search(/[^\s\u00a0]/),-1==b&&(b=a.length));d=d||0;for(e=e||0;;){var g=a.indexOf("\t",d);if(0>g||g>=b)return e+(b-d);
 e+=g-d;e+=c-e%c;d=g+1}}function v(a,b){for(var c=0;c<a.length;++c)if(a[c]==b)return c;return-1}function J(a,b,c){for(var d=0,e=0;;){var g=a.indexOf("\t",d);-1==g&&(g=a.length);var h=g-d;if(g==a.length||e+h>=b)return d+Math.min(h,b-e);e+=g-d;e+=c-e%c;d=g+1;if(e>=b)return d}}function G(a){for(;nc.length<=a;)nc.push(z(nc)+" ");return nc[a]}function z(a){return a[a.length-1]}function Fa(a,b){for(var c=[],d=0;d<a.length;d++)c[d]=b(a[d],d);return c}function oc(a,b,c){for(var d=0,e=c(b);d<a.length&&c(a[d])<=
@@ -326,7 +326,7 @@ T.defineMIME("text/plain","null");T.defineExtension=function(a,b){T.prototype[a]
 a.form;e=g.submit;try{var h=g.submit=function(){c();g.submit=e;g.submit();g.submit=h}}catch(l){}}b.finishInit=function(d){d.save=c;d.getTextArea=function(){return a};d.toTextArea=function(){d.toTextArea=isNaN;c();a.parentNode.removeChild(d.getWrapperElement());a.style.display="";a.form&&(R(a.form,"submit",c),b.leaveSubmitMethodAlone||"function"!=typeof a.form.submit||(a.form.submit=e))}};a.style.display="none";var f=T(function(b){return a.parentNode.insertBefore(b,a.nextSibling)},b);return f};(function(a){a.off=
 R;a.on=E;a.wheelEventPixels=Eg;a.Doc=la;a.splitLines=Wd;a.countColumn=I;a.findColumn=J;a.isWordChar=C;a.Pass=Qc;a.signal=X;a.Line=rb;a.changeEnd=Oa;a.scrollbarModel=Ve;a.Pos=w;a.cmpPos=K;a.modes=Xc;a.mimeModes=ib;a.resolveMode=qc;a.getMode=Yc;a.modeExtensions=jb;a.extendMode=jg;a.copyState=Ta;a.startState=ae;a.innerMode=Zc;a.commands=ac;a.keyMap=$b;a.keyName=Gf;a.isModifierKey=Df;a.lookupKey=xb;a.normalizeKeyMap=Qg;a.StringStream=Y;a.SharedTextMarker=Zb;a.TextMarker=Qa;a.LineWidget=Yb;a.e_preventDefault=
 ja;a.e_stopPropagation=Zd;a.e_stop=Db;a.addClass=B;a.contains=A;a.rmClass=ab;a.keyNames=Ra})(T);T.version="5.57.0";return T});
-(function(Q){"object"==typeof exports&&"object"==typeof module?Q(require("../../lib/codemirror")):"function"==typeof define&&define.amd?define(["../../lib/codemirror"],Q):Q(CodeMirror)})(function(Q){var da={autoSelfClosers:{area:!0,base:!0,br:!0,col:!0,command:!0,embed:!0,frame:!0,hr:!0,img:!0,input:!0,keygen:!0,link:!0,meta:!0,param:!0,source:!0,track:!0,wbr:!0,menuitem:!0},implicitlyClosed:{dd:!0,li:!0,optgroup:!0,option:!0,p:!0,rp:!0,rt:!0,tbody:!0,td:!0,tfoot:!0,th:!0,tr:!0},contextGrabbers:{dd:{dd:!0,
+(function(Q){"object"==typeof exports&&"object"==typeof module?Q(require("../../lib/codemirror")):"function"==__define_disabled__&&define.amd?define(["../../lib/codemirror"],Q):Q(CodeMirror)})(function(Q){var da={autoSelfClosers:{area:!0,base:!0,br:!0,col:!0,command:!0,embed:!0,frame:!0,hr:!0,img:!0,input:!0,keygen:!0,link:!0,meta:!0,param:!0,source:!0,track:!0,wbr:!0,menuitem:!0},implicitlyClosed:{dd:!0,li:!0,optgroup:!0,option:!0,p:!0,rp:!0,rt:!0,tbody:!0,td:!0,tfoot:!0,th:!0,tr:!0},contextGrabbers:{dd:{dd:!0,
 dt:!0},dt:{dd:!0,dt:!0},li:{li:!0},option:{option:!0,optgroup:!0},optgroup:{optgroup:!0},p:{address:!0,article:!0,aside:!0,blockquote:!0,dir:!0,div:!0,dl:!0,fieldset:!0,footer:!0,form:!0,h1:!0,h2:!0,h3:!0,h4:!0,h5:!0,h6:!0,header:!0,hgroup:!0,hr:!0,menu:!0,nav:!0,ol:!0,p:!0,pre:!0,section:!0,table:!0,ul:!0},rp:{rp:!0,rt:!0},rt:{rp:!0,rt:!0},tbody:{tbody:!0,tfoot:!0},td:{td:!0,th:!0},tfoot:{tbody:!0},th:{td:!0,th:!0},thead:{tbody:!0,tfoot:!0},tr:{tr:!0}},doNotIndent:{pre:!0},allowUnquoted:!0,allowMissing:!0,
 caseFold:!0},Z={autoSelfClosers:{},implicitlyClosed:{},contextGrabbers:{},doNotIndent:{},allowUnquoted:!1,allowMissing:!1,allowMissingTagName:!1,caseFold:!1};Q.defineMode("xml",function(r,y){function A(k,t){function C(f){t.tokenize=f;return f(k,t)}var H=k.next();if("\x3c"==H){if(k.eat("!"))return k.eat("[")?k.match("CDATA[")?C(n("atom","]]\x3e")):null:k.match("--")?C(n("comment","--\x3e")):k.match("DOCTYPE",!0,!0)?(k.eatWhile(/[\w\._\-]/),C(f(1))):null;if(k.eat("?"))return k.eatWhile(/[\w\._\-]/),
 t.tokenize=n("meta","?\x3e"),"meta";Ka=k.eat("/")?"closeTag":"openTag";t.tokenize=m;return"tag bracket"}if("\x26"==H)return(k.eat("#")?k.eat("x")?k.eatWhile(/[a-fA-F\d]/)&&k.eat(";"):k.eatWhile(/[\d]/)&&k.eat(";"):k.eatWhile(/[\w\.\-:]/)&&k.eat(";"))?"atom":"error";k.eatWhile(/[^&<]/);return null}function m(f,k){var t=f.next();if("\x3e"==t||"/"==t&&f.eat("\x3e"))return k.tokenize=A,Ka="\x3e"==t?"endTag":"selfcloseTag","tag bracket";if("\x3d"==t)return Ka="equals",null;if("\x3c"==t)return k.tokenize=
@@ -421,5 +421,5 @@ n=this._create_token(k.TAG_OPEN,n));return n};A.prototype._read_open_handlebars=
 f||"/"===f&&"\x3e"===this._input.peek(1))?(n=this._input.next(),"/"===f&&(n+=this._input.next()),n=this._create_token(k.TAG_CLOSE,n)):"{"===m.text[0]&&"}"===f&&"}"===this._input.peek(1)&&(this._input.next(),this._input.next(),n=this._create_token(k.TAG_CLOSE,"}}")));return n};A.prototype._read_attribute=function(f,m,n){var r=null,y="";n&&"\x3c"===n.text[0]&&("\x3d"===f?r=this._create_token(k.EQUALS,this._input.next()):'"'===f||"'"===f?(m=this._input.next(),m='"'===f?m+this.__patterns.double_quote.read():
 m+this.__patterns.single_quote.read(),r=this._create_token(k.VALUE,m)):(y=this.__patterns.attribute.read())&&(r=m.type===k.EQUALS?this._create_token(k.VALUE,y):this._create_token(k.ATTRIBUTE,y)));return r};A.prototype._is_content_unformatted=function(f){return-1===this._options.void_elements.indexOf(f)&&(-1!==this._options.content_unformatted.indexOf(f)||-1!==this._options.unformatted.indexOf(f))};A.prototype._read_raw_content=function(f,m,n){var r="";if(n&&"{"===n.text[0])r=this.__patterns.handlebars_raw_close.read();
 else if(m.type===k.TAG_CLOSE&&"\x3c"===m.opened.text[0]&&"/"!==m.text[0])if(m=m.opened.text.substr(1).toLowerCase(),"script"===m||"style"===m){if(f=this._read_comment_or_cdata(f))return f.type=k.TEXT,f;r=this._input.readUntil(new RegExp("\x3c/"+m+"[\\n\\r\\t ]*?\x3e","ig"))}else this._is_content_unformatted(m)&&(r=this._input.readUntil(new RegExp("\x3c/"+m+"[\\n\\r\\t ]*?\x3e","ig")));return r?this._create_token(k.TEXT,r):null};A.prototype._read_content_word=function(f){var m="";this._options.unformatted_content_delimiter&&
-f===this._options.unformatted_content_delimiter[0]&&(m=this.__patterns.unformatted_content_delimiter.read());m||(m=this.__patterns.word.read());if(m)return this._create_token(k.TEXT,m)};r.exports.Tokenizer=A;r.exports.TOKEN=k}]);if("function"===typeof define&&define.amd)define(["require","./beautify","./beautify-css"],function(r){var y=r("./beautify"),A=r("./beautify-css");return{html_beautify:function(m,r){return Q(m,r,y.js_beautify,A.css_beautify)}}});else if("undefined"!==typeof exports){var da=
+f===this._options.unformatted_content_delimiter[0]&&(m=this.__patterns.unformatted_content_delimiter.read());m||(m=this.__patterns.word.read());if(m)return this._create_token(k.TEXT,m)};r.exports.Tokenizer=A;r.exports.TOKEN=k}]);if("function"===__define_disabled__&&define.amd)define(["require","./beautify","./beautify-css"],function(r){var y=r("./beautify"),A=r("./beautify-css");return{html_beautify:function(m,r){return Q(m,r,y.js_beautify,A.css_beautify)}}});else if("undefined"!==typeof exports){var da=
 require("./beautify.js"),Z=require("./beautify-css.js");exports.html_beautify=function(r,y){return Q(r,y,da.js_beautify,Z.css_beautify)}}else"undefined"!==typeof window?window.html_beautify=function(r,y){return Q(r,y,window.js_beautify,window.css_beautify)}:"undefined"!==typeof global&&(global.html_beautify=function(r,y){return Q(r,y,global.js_beautify,global.css_beautify)})})();
\ No newline at end of file
```

wincent commented:
So there's a bug in there:

`typeof define` got changed to `__define_disabled__`

You want it to be `typeof __define_disable__`.

I hereby withdraw my approval! 😉

julien commented:
Yes, that probably because I was using  `path.replaceWithSourceString('__define_disabled__');` which I now updated.


Here's an updated diff 

```diff
diff --git a/ckeditor/plugins/codemirror/vendors/vendors.js b/ckeditor/plugins/codemirror/vendors/vendors.js
index dd1302231..205f273f8 100644
--- a/ckeditor/plugins/codemirror/vendors/vendors.js
+++ b/ckeditor/plugins/codemirror/vendors/vendors.js
@@ -1,4 +1,4 @@
-﻿(function(Q,da){"object"===typeof exports&&"undefined"!==typeof module?module.exports=da():"function"===typeof define&&define.amd?define(da):(Q=Q||self,Q.CodeMirror=da())})(this,function(){function Q(a){return new RegExp("(^|\\s)"+a+"(?:$|\\s)\\s*")}function da(a){for(var b=a.childNodes.length;0<b;--b)a.removeChild(a.firstChild);return a}function Z(a,b){return da(a).appendChild(b)}function r(a,b,c,d){a=document.createElement(a);c&&(a.className=c);d&&(a.style.cssText=d);if("string"==typeof b)a.appendChild(document.createTextNode(b));
+﻿(function(Q,da){"object"===typeof exports&&"undefined"!==typeof module?module.exports=da():"function"===typeof __define_disabled__&&define.amd?define(da):(Q=Q||self,Q.CodeMirror=da())})(this,function(){function Q(a){return new RegExp("(^|\\s)"+a+"(?:$|\\s)\\s*")}function da(a){for(var b=a.childNodes.length;0<b;--b)a.removeChild(a.firstChild);return a}function Z(a,b){return da(a).appendChild(b)}function r(a,b,c,d){a=document.createElement(a);c&&(a.className=c);d&&(a.style.cssText=d);if("string"==typeof b)a.appendChild(document.createTextNode(b));
 else if(b)for(c=0;c<b.length;++c)a.appendChild(b[c]);return a}function y(a,b,c,d){a=r(a,b,c,d);a.setAttribute("role","presentation");return a}function A(a,b){3==b.nodeType&&(b=b.parentNode);if(a.contains)return a.contains(b);do if(11==b.nodeType&&(b=b.host),b==a)return!0;while(b=b.parentNode)}function m(){var a;try{a=document.activeElement}catch(b){a=document.body||null}for(;a&&a.shadowRoot&&a.shadowRoot.activeElement;)a=a.shadowRoot.activeElement;return a}function B(a,b){var c=a.className;Q(b).test(c)||
 (a.className+=(c?" ":"")+b)}function n(a,b){for(var c=a.split(" "),d=0;d<c.length;d++)c[d]&&!Q(c[d]).test(b)&&(b+=" "+c[d]);return b}function f(a){var b=Array.prototype.slice.call(arguments,1);return function(){return a.apply(null,b)}}function k(a,b,c){b||(b={});for(var d in a)!a.hasOwnProperty(d)||!1===c&&b.hasOwnProperty(d)||(b[d]=a[d]);return b}function I(a,b,c,d,e){null==b&&(b=a.search(/[^\s\u00a0]/),-1==b&&(b=a.length));d=d||0;for(e=e||0;;){var g=a.indexOf("\t",d);if(0>g||g>=b)return e+(b-d);
 e+=g-d;e+=c-e%c;d=g+1}}function v(a,b){for(var c=0;c<a.length;++c)if(a[c]==b)return c;return-1}function J(a,b,c){for(var d=0,e=0;;){var g=a.indexOf("\t",d);-1==g&&(g=a.length);var h=g-d;if(g==a.length||e+h>=b)return d+Math.min(h,b-e);e+=g-d;e+=c-e%c;d=g+1;if(e>=b)return d}}function G(a){for(;nc.length<=a;)nc.push(z(nc)+" ");return nc[a]}function z(a){return a[a.length-1]}function Fa(a,b){for(var c=[],d=0;d<a.length;d++)c[d]=b(a[d],d);return c}function oc(a,b,c){for(var d=0,e=c(b);d<a.length&&c(a[d])<=
@@ -326,7 +326,7 @@ T.defineMIME("text/plain","null");T.defineExtension=function(a,b){T.prototype[a]
 a.form;e=g.submit;try{var h=g.submit=function(){c();g.submit=e;g.submit();g.submit=h}}catch(l){}}b.finishInit=function(d){d.save=c;d.getTextArea=function(){return a};d.toTextArea=function(){d.toTextArea=isNaN;c();a.parentNode.removeChild(d.getWrapperElement());a.style.display="";a.form&&(R(a.form,"submit",c),b.leaveSubmitMethodAlone||"function"!=typeof a.form.submit||(a.form.submit=e))}};a.style.display="none";var f=T(function(b){return a.parentNode.insertBefore(b,a.nextSibling)},b);return f};(function(a){a.off=
 R;a.on=E;a.wheelEventPixels=Eg;a.Doc=la;a.splitLines=Wd;a.countColumn=I;a.findColumn=J;a.isWordChar=C;a.Pass=Qc;a.signal=X;a.Line=rb;a.changeEnd=Oa;a.scrollbarModel=Ve;a.Pos=w;a.cmpPos=K;a.modes=Xc;a.mimeModes=ib;a.resolveMode=qc;a.getMode=Yc;a.modeExtensions=jb;a.extendMode=jg;a.copyState=Ta;a.startState=ae;a.innerMode=Zc;a.commands=ac;a.keyMap=$b;a.keyName=Gf;a.isModifierKey=Df;a.lookupKey=xb;a.normalizeKeyMap=Qg;a.StringStream=Y;a.SharedTextMarker=Zb;a.TextMarker=Qa;a.LineWidget=Yb;a.e_preventDefault=
 ja;a.e_stopPropagation=Zd;a.e_stop=Db;a.addClass=B;a.contains=A;a.rmClass=ab;a.keyNames=Ra})(T);T.version="5.57.0";return T});
-(function(Q){"object"==typeof exports&&"object"==typeof module?Q(require("../../lib/codemirror")):"function"==typeof define&&define.amd?define(["../../lib/codemirror"],Q):Q(CodeMirror)})(function(Q){var da={autoSelfClosers:{area:!0,base:!0,br:!0,col:!0,command:!0,embed:!0,frame:!0,hr:!0,img:!0,input:!0,keygen:!0,link:!0,meta:!0,param:!0,source:!0,track:!0,wbr:!0,menuitem:!0},implicitlyClosed:{dd:!0,li:!0,optgroup:!0,option:!0,p:!0,rp:!0,rt:!0,tbody:!0,td:!0,tfoot:!0,th:!0,tr:!0},contextGrabbers:{dd:{dd:!0,
+(function(Q){"object"==typeof exports&&"object"==typeof module?Q(require("../../lib/codemirror")):"function"==typeof __define_disabled__&&define.amd?define(["../../lib/codemirror"],Q):Q(CodeMirror)})(function(Q){var da={autoSelfClosers:{area:!0,base:!0,br:!0,col:!0,command:!0,embed:!0,frame:!0,hr:!0,img:!0,input:!0,keygen:!0,link:!0,meta:!0,param:!0,source:!0,track:!0,wbr:!0,menuitem:!0},implicitlyClosed:{dd:!0,li:!0,optgroup:!0,option:!0,p:!0,rp:!0,rt:!0,tbody:!0,td:!0,tfoot:!0,th:!0,tr:!0},contextGrabbers:{dd:{dd:!0,
 dt:!0},dt:{dd:!0,dt:!0},li:{li:!0},option:{option:!0,optgroup:!0},optgroup:{optgroup:!0},p:{address:!0,article:!0,aside:!0,blockquote:!0,dir:!0,div:!0,dl:!0,fieldset:!0,footer:!0,form:!0,h1:!0,h2:!0,h3:!0,h4:!0,h5:!0,h6:!0,header:!0,hgroup:!0,hr:!0,menu:!0,nav:!0,ol:!0,p:!0,pre:!0,section:!0,table:!0,ul:!0},rp:{rp:!0,rt:!0},rt:{rp:!0,rt:!0},tbody:{tbody:!0,tfoot:!0},td:{td:!0,th:!0},tfoot:{tbody:!0},th:{td:!0,th:!0},thead:{tbody:!0,tfoot:!0},tr:{tr:!0}},doNotIndent:{pre:!0},allowUnquoted:!0,allowMissing:!0,
 caseFold:!0},Z={autoSelfClosers:{},implicitlyClosed:{},contextGrabbers:{},doNotIndent:{},allowUnquoted:!1,allowMissing:!1,allowMissingTagName:!1,caseFold:!1};Q.defineMode("xml",function(r,y){function A(k,t){function C(f){t.tokenize=f;return f(k,t)}var H=k.next();if("\x3c"==H){if(k.eat("!"))return k.eat("[")?k.match("CDATA[")?C(n("atom","]]\x3e")):null:k.match("--")?C(n("comment","--\x3e")):k.match("DOCTYPE",!0,!0)?(k.eatWhile(/[\w\._\-]/),C(f(1))):null;if(k.eat("?"))return k.eatWhile(/[\w\._\-]/),
 t.tokenize=n("meta","?\x3e"),"meta";Ka=k.eat("/")?"closeTag":"openTag";t.tokenize=m;return"tag bracket"}if("\x26"==H)return(k.eat("#")?k.eat("x")?k.eatWhile(/[a-fA-F\d]/)&&k.eat(";"):k.eatWhile(/[\d]/)&&k.eat(";"):k.eatWhile(/[\w\.\-:]/)&&k.eat(";"))?"atom":"error";k.eatWhile(/[^&<]/);return null}function m(f,k){var t=f.next();if("\x3e"==t||"/"==t&&f.eat("\x3e"))return k.tokenize=A,Ka="\x3e"==t?"endTag":"selfcloseTag","tag bracket";if("\x3d"==t)return Ka="equals",null;if("\x3c"==t)return k.tokenize=
@@ -421,5 +421,5 @@ n=this._create_token(k.TAG_OPEN,n));return n};A.prototype._read_open_handlebars=
 f||"/"===f&&"\x3e"===this._input.peek(1))?(n=this._input.next(),"/"===f&&(n+=this._input.next()),n=this._create_token(k.TAG_CLOSE,n)):"{"===m.text[0]&&"}"===f&&"}"===this._input.peek(1)&&(this._input.next(),this._input.next(),n=this._create_token(k.TAG_CLOSE,"}}")));return n};A.prototype._read_attribute=function(f,m,n){var r=null,y="";n&&"\x3c"===n.text[0]&&("\x3d"===f?r=this._create_token(k.EQUALS,this._input.next()):'"'===f||"'"===f?(m=this._input.next(),m='"'===f?m+this.__patterns.double_quote.read():
 m+this.__patterns.single_quote.read(),r=this._create_token(k.VALUE,m)):(y=this.__patterns.attribute.read())&&(r=m.type===k.EQUALS?this._create_token(k.VALUE,y):this._create_token(k.ATTRIBUTE,y)));return r};A.prototype._is_content_unformatted=function(f){return-1===this._options.void_elements.indexOf(f)&&(-1!==this._options.content_unformatted.indexOf(f)||-1!==this._options.unformatted.indexOf(f))};A.prototype._read_raw_content=function(f,m,n){var r="";if(n&&"{"===n.text[0])r=this.__patterns.handlebars_raw_close.read();
 else if(m.type===k.TAG_CLOSE&&"\x3c"===m.opened.text[0]&&"/"!==m.text[0])if(m=m.opened.text.substr(1).toLowerCase(),"script"===m||"style"===m){if(f=this._read_comment_or_cdata(f))return f.type=k.TEXT,f;r=this._input.readUntil(new RegExp("\x3c/"+m+"[\\n\\r\\t ]*?\x3e","ig"))}else this._is_content_unformatted(m)&&(r=this._input.readUntil(new RegExp("\x3c/"+m+"[\\n\\r\\t ]*?\x3e","ig")));return r?this._create_token(k.TEXT,r):null};A.prototype._read_content_word=function(f){var m="";this._options.unformatted_content_delimiter&&
-f===this._options.unformatted_content_delimiter[0]&&(m=this.__patterns.unformatted_content_delimiter.read());m||(m=this.__patterns.word.read());if(m)return this._create_token(k.TEXT,m)};r.exports.Tokenizer=A;r.exports.TOKEN=k}]);if("function"===typeof define&&define.amd)define(["require","./beautify","./beautify-css"],function(r){var y=r("./beautify"),A=r("./beautify-css");return{html_beautify:function(m,r){return Q(m,r,y.js_beautify,A.css_beautify)}}});else if("undefined"!==typeof exports){var da=
+f===this._options.unformatted_content_delimiter[0]&&(m=this.__patterns.unformatted_content_delimiter.read());m||(m=this.__patterns.word.read());if(m)return this._create_token(k.TEXT,m)};r.exports.Tokenizer=A;r.exports.TOKEN=k}]);if("function"===typeof __define_disabled__&&define.amd)define(["require","./beautify","./beautify-css"],function(r){var y=r("./beautify"),A=r("./beautify-css");return{html_beautify:function(m,r){return Q(m,r,y.js_beautify,A.css_beautify)}}});else if("undefined"!==typeof exports){var da=
 require("./beautify.js"),Z=require("./beautify-css.js");exports.html_beautify=function(r,y){return Q(r,y,da.js_beautify,Z.css_beautify)}}else"undefined"!==typeof window?window.html_beautify=function(r,y){return Q(r,y,window.js_beautify,window.css_beautify)}:"undefined"!==typeof global&&(global.html_beautify=function(r,y){return Q(r,y,global.js_beautify,global.css_beautify)})})();
\ No newline at end of file
```

wincent commented:
lg2m

carloslancha reviewed (approved):
lgtm
